src/blocks/calendar.rs

```diff
@@ -0,0 +1,396 @@
+
+mod auth;
+mod caldav;
+
+use self::auth::{AuthorizeUrl, OAuth2Flow, TokenStore};
+use self::caldav::Event;
+
+use super::prelude::*;
+
+use std::path::Path;
+
+use caldav::CalDavClient;
+
+#[derive(Deserialize, Debug)]
+pub struct BasicAuthConfig {
+    username: Option<String>,
+    password: Option<String>,
+}
+
+#[derive(Deserialize, Debug, SmartDefault)]
+#[serde(deny_unknown_fields, default)]
+pub struct OAuth2Config {
+    pub client_id: Option<String>,
+    pub client_secret: Option<String>,
+    pub auth_url: String,
+    pub token_url: String,
+    #[default("~/.config/i3status-rust/calendar.auth_token".into())]
+    pub auth_token: ShellString,
+    #[default(8080)]
+    pub redirect_port: u16,
+    pub scopes: Vec<Scope>,
+}
+
+#[derive(Deserialize, Default, Debug)]
+#[serde(tag = "type", rename_all = "lowercase")]
+pub enum AuthConfig {
+    #[default]
+    Unauthenticated,
+    Basic(BasicAuthConfig),
+    OAuth2(OAuth2Config),
+}
+
+#[derive(Deserialize, Debug, SmartDefault)]
+#[serde(deny_unknown_fields, default)]
+pub struct Config {
+    pub next_event_format: FormatConfig,
+    pub ongoing_event_format: FormatConfig,
+    pub no_events_format: FormatConfig,
+    pub redirect_format: FormatConfig,
+    #[default(60.into())]
+    pub interval: Seconds,
+    pub url: String,
+    pub auth: AuthConfig,
+    pub calendars: Vec<String>,
+    #[default(48)]
+    pub events_within_hours: u64,
+    #[default(300.into())]
+    pub warning_threshold: Seconds,
+    #[default("xdg-open".into())]
+    pub browser_cmd: ShellString,
+}
+
+pub async fn run(config: &Config, api: &CommonApi) -> Result<()> {
+    let next_event_format = config
+        .next_event_format
+        .with_default(" $icon $start.datetime(f:'%a %H:%M') $summary ")?;
+    let ongoing_event_format = config
+        .ongoing_event_format
+        .with_default(" $icon $summary (ends at $end.datetime(f:'%H:%M')) ")?;
+    let no_events_format = config.no_events_format.with_default(" $icon ")?;
+    let redirect_format = config
+        .redirect_format
+        .with_default(" $icon Check your web browser ")?;
+
+    api.set_default_actions(&[(MouseButton::Left, None, "open_link")])?;
+
+    let mut client = caldav_client(config).await?;
+
+    let mut timer = config.interval.timer();
+
+    let mut actions = api.get_actions()?;
+
+    loop {
+        let mut widget = Widget::new().with_format(no_events_format.clone());
+        widget.set_values(map! {
+            "icon" => Value::icon("calendar"),
+        });
+
+        let mut next_event = None;
+        let mut retries = 0;
+        while retries <= 1 {
+            let next_event_result = get_next_event(config, &mut client).await;
+            match next_event_result {
+                Ok(event) => {
+                    next_event = event;
+                    break;
+                }
+                Err(err) => {
+                    if let CalendarError::AuthRequired = err {
```

Comment:
You should surface the other errors that could have happended for example `CalendarError::Http` could be raised for using the wrong url and you get a 404 error.

```diff
diff --git a/src/blocks/calendar.rs b/src/blocks/calendar.rs
index efb5aa114..6b12f7b43 100644
--- a/src/blocks/calendar.rs
+++ b/src/blocks/calendar.rs
@@ -131,6 +131,7 @@ use self::caldav::Event;
 use super::prelude::*;
 
 use std::path::Path;
+use std::sync::Arc;
 
 use caldav::CalDavClient;
 
@@ -218,8 +219,8 @@ pub async fn run(config: &Config, api: &CommonApi) -> Result<()> {
                     next_event = event;
                     break;
                 }
-                Err(err) => {
-                    if let CalendarError::AuthRequired = err {
+                Err(err) => match err {
+                    CalendarError::AuthRequired => {
                         let authorization =
                             client.authorize().await.error("Authorization failed")?;
                         match &authorization {
@@ -239,7 +240,13 @@ pub async fn run(config: &Config, api: &CommonApi) -> Result<()> {
                             }
                         }
                     }
-                }
+                    e => {
+                        return Err(Error {
+                            message: None,
+                            cause: Some(Arc::new(e)),
+                        })
+                    }
+                },
             };
             retries += 1;
         }
```